src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3542,24 +3566,85 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
             return (method, results, shouldReturnNotNull);
         }
 
-        private ImmutableArray<VisitArgumentResult> VisitArgumentsEvaluate(ImmutableArray<BoundExpression> arguments, ImmutableArray<RefKind> refKindsOpt, ImmutableArray<ParameterSymbol> parameters, ImmutableArray<int> argsToParamsOpt, bool expanded)
+#nullable enable
+        private (ImmutableArray<VisitArgumentResult> results, ImmutableArray<int> allArgsToParamsOpt, ImmutableArray<RefKind> allRefKindsOpt) VisitArgumentsEvaluate(
+            SyntaxNode syntax,
+            ImmutableArray<BoundExpression> arguments,
+            ImmutableArray<RefKind> refKindsOpt,
+            ImmutableArray<ParameterSymbol> parametersOpt,
+            ImmutableArray<int> argsToParamsOpt,
+            bool expanded)
         {
             Debug.Assert(!IsConditionalState);
             int n = arguments.Length;
-            if (n == 0)
+            if (n == 0 && parametersOpt.IsDefaultOrEmpty)
             {
-                return ImmutableArray<VisitArgumentResult>.Empty;
+                return (ImmutableArray<VisitArgumentResult>.Empty, argsToParamsOpt, refKindsOpt);
             }
-            var builder = ArrayBuilder<VisitArgumentResult>.GetInstance(n);
+
+            var visitedParameters = PooledHashSet<ParameterSymbol>.GetInstance();
```

Comment:
>() [](start = 78, length = 2)

can use `n` as size hint?

Reply:
This seems reasonable, as long as it doesn't cause a bad behavior when we have a call with a large params array.

---
In reply to: [329248589](https://github.com/dotnet/roslyn/pull/38788#discussion_r329248589) [](ancestors = 329248589)

Reply:
Scratch that, PooledHashSet.GetInstance does not actually accept a size hint.

---
In reply to: [329263552](https://github.com/dotnet/roslyn/pull/38788#discussion_r329263552) [](ancestors = 329263552,329248589)

Reply:
Ah, good point

---
In reply to: [329263995](https://github.com/dotnet/roslyn/pull/38788#discussion_r329263995) [](ancestors = 329263995,329263552,329248589)